File: poj/3101_Astronomy_拓展欧几里得.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

unsigned long long p[3];

unsigned long long* gcdEx(unsigned long long a, unsigned long long b, unsigned long long *p) {
    if (b == 0) {
        p[0] = 1, p[1] = 0, p[2] = a;
        return p;
    } else {
        p = gcdEx(b, b%a, p);
        unsigned long long t = p[0];
        p[0] = p[1];
        p[1] = t - a/b*p[1];
        return p;
    }
}

void swap(unsigned long long &a, unsigned long long &b) {
    if (a != b) {
        a ^= b ^= a ^= b;
    }
}

unsigned long long gcd(unsigned long long a, unsigned long long b) {
    if (a == 0 || b == 0) return 0;
    while (a && b) b%=a^=b^=a^=b;
    return a+b;
}

void solve() {
    unsigned long long n;

    scanf("%llu", &n);

    // ans  = ans1 / ans2;
    unsigned long long ans1, ans2;

    unsigned long long a, b;

    scanf("%llu", &a);
    scanf("%llu", &b);
    if (a > b) {
        swap(a, b);
    }
    ans1 = a * b;
    ans2 = 2 * (b - a);
    unsigned long long g = gcd(ans1, ans2);
    if (g != 0) {
        ans1 /= g;
        ans2 /= g;
    } else {
        ans1 = 0;
        ans2 = 1;
    }


    for (unsigned long long i = 2; i < n; ++ i) {
        scanf("%llu", &b);

        if (a > b) {
            swap(a, b);
        }
        unsigned long long newans1, newans2;
        newans1 = a * b;
        newans2 = 2 * (b - a);

        unsigned long long g = gcd(newans1, newans2);
        if (g != 0) {
            newans1 /= g;
            newans2 /= g;
        } else {
            continue;
        }

        if (ans1 == 0) {
            ans1 = newans1;
            ans2 = newans2;
            continue;
        }

        g = gcd(ans1 * newans2, newans1 * ans2);
        ans1 *= newans1 * ans2 / g;
        g = gcd(ans1, ans2);
        ans1 /= g;
        ans2 /= g;
    }

    printf("%llu %llu", ans1, ans2);
}

void test() {
    gcdEx(8, 5, p);
    printf("%llu %llu %llu\n", p[0], p[1], p[2]);
}

int main() {
    solve();

    return 0;
}

File: poj/3101_Astronomy_拓展欧几里得.java
import java.math.BigInteger;
import java.io.*;
import java.util.StringTokenizer;

// Rename to "Main" when submit to POJ
public class _3101_Astronomy {
    private static BigInteger Zero = new BigInteger("0");
    private static BigInteger One = new BigInteger("1");
    private static BigInteger Two = new BigInteger("2");

    public static void main(String[] args) throws IOException {
        BufferedReader f = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter out = new PrintWriter(System.out);

        StringTokenizer st = new StringTokenizer(f.readLine());
        int n = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(f.readLine());

        BigInteger a = new BigInteger(st.nextToken());
        BigInteger b = new BigInteger(st.nextToken());

        BigInteger ans1 = a.multiply(b);
        BigInteger ans2 = a.subtract(b).abs();//.multiply(Two);
        BigInteger g = ans1.gcd(ans2);
        if (!ans2.equals(Zero)) {
            ans1 = ans1.divide(g);
            ans2 = ans2.divide(g);
        } else {
            ans1 = Zero;
            ans2 = One;
        }

        for (int i = 2; i < n; ++ i) {
            b = new BigInteger(st.nextToken());

            BigInteger newans1 = a.multiply(b);
            BigInteger newans2 = a.subtract(b).abs();//.multiply(Two);
            g = newans1.gcd(newans2);
            if (!newans2.equals(Zero)) {
                newans1 = newans1.divide(g);
                newans2 = newans2.divide(g);
            } else {
                continue;
            }

            if (ans1.equals(Zero)) {
                ans1 = newans1;
                ans2 = newans2;
                continue;
            }

            g = ans1.multiply(newans2).gcd(newans1.multiply(ans2));
            ans1 = ans1.multiply(newans1).multiply(ans2).divide(g);
            g = ans1.gcd(ans2);
            ans1 = ans1.divide(g);
            ans2 = ans2.divide(g);
        }
        // Optimize: See the commented //.multiply(Two) in above lines.
        if (ans1.mod(Two).equals(Zero)) {
            ans1 = ans1.divide(Two);
        } else {
            ans2 = ans2.multiply(Two);
        }

        System.out.print(ans1);
        System.out.print(" ");
        System.out.println(ans2);

        f.close();
        out.close(); // close the output file
        System.exit(0); // don't omit this!
    }
}

